File: LeetCode/April Challenge/27th Day/Day_27.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m=matrix.size();
        if(m==0)
            return 0;
        int n=matrix[0].size();        
        int sq[m+1][n+1];
        int res=0;
        for(int i=0;i<=m;i++)
        {
            for(int j=0; j<=n;j++)
            {
                if(i==0 || j==0 || matrix[i-1][j-1]=='0')
                    sq[i][j]=0;
                else
                    sq[i][j]=1+ min(sq[i-1][j-1], min( sq[i-1][j], sq[i][j-1]));
                
                res=max(res, sq[i][j]);
            }
        }
        return res*res;      
    }
};

File: LeetCode/April Challenge/27th Day/Day_27.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if len(matrix) == 0:
            return 0
        n, m = len(matrix), len(matrix[0])
        dp = [[0 for j in range(m + 1)] for i in range(n + 1)]
        max_area = 0
        for i in range(n + 1):
            for j in range(m + 1):
                if i == 0 or j == 0 or matrix[i - 1][j - 1] == '0':
                    dp[i][j] = 0
                else:
                    dp[i][j] = min(min(dp[i][j - 1], dp[i - 1][j]), dp[i - 1][j - 1]) + 1
                    max_area = max(max_area, dp[i][j])
        return max_area * max_area
    
               
